metanode/partition_file_stats.go

```diff
@@ -0,0 +1,215 @@
+		return "<1M"
+	case LessThan16M:
+		return "<16M"
+	case LessThan32M:
+		return "<32M"
+	case LessThan64M:
+		return "<64M"
+	case LessThan128M:
+		return "<128M"
+	case LessThan256M:
+		return "<256M"
+	case BiggerThan256M:
+		return ">256M"
+	default:
+		return "unknown"
+	}
+}
+
+func (mp *metaPartition) setMetrics(fileRange []int64) {
+	for i, val := range fileRange {
+		labels := map[string]string{
+			"partid":    fmt.Sprintf("%d", mp.config.PartitionId),
+			"volName":   mp.config.VolName,
+			"sizeRange": toString(FileSizeRange(i)),
+		}
+		exporter.NewGauge("fileStats").SetWithLabels(float64(val), labels)
+	}
+}
+
+func (mp *metaPartition) fileStats(rootDir string) {
+	var err error
+	fileRange := make([]int64, MaxRangeType)
+	defer func() {
+		if err != nil {
+			log.LogErrorf("fileStats failed: %v", err.Error())
+			return
+		}
+		mp.setMetrics(fileRange)
+	}()
+	filename := path.Join(rootDir, inodeFile)
+	if _, err = os.Stat(filename); err != nil {
+		err = errors.NewErrorf("Stat: %s", err.Error())
+		return
+	}
+	fp, err := os.OpenFile(filename, os.O_RDONLY, 0644)
+	if err != nil {
+		err = errors.NewErrorf("OpenFile: %s", err.Error())
+		return
+	}
+	defer fp.Close()
+	reader := bufio.NewReaderSize(fp, 4*1024*1024)
+	inoBuf := make([]byte, 4)
+
+	var count uint64
+	for {
+		inoBuf = inoBuf[:4]
+		// first read length
+		_, err = io.ReadFull(reader, inoBuf)
+		if err != nil {
+			if err == io.EOF {
+				err = nil
+				return
+			}
+			err = errors.NewErrorf("ReadHeader: %s", err.Error())
+			return
+		}
+		length := binary.BigEndian.Uint32(inoBuf)
+
+		// next read body
+		if uint32(cap(inoBuf)) >= length {
+			inoBuf = inoBuf[:length]
+		} else {
+			inoBuf = make([]byte, length)
+		}
+		_, err = io.ReadFull(reader, inoBuf)
+		if err != nil {
+			err = errors.NewErrorf("ReadBody: %s", err.Error())
+			return
+		}
+		ino := NewInode(0, 0)
+		if err = ino.Unmarshal(inoBuf); err != nil {
+			err = errors.NewErrorf("Unmarshal: %s", err.Error())
+			return
+		}
+
+		if ino.NLink > 0 && proto.IsRegular(ino.Type) {
+			if 0 <= ino.Size && ino.Size < Size1K {
+				fileRange[LessThan1K] += 1
+			} else if Size1K <= ino.Size && ino.Size < Size1M {
+				fileRange[LessThan1M] += 1
+			} else if Size1M <= ino.Size && ino.Size < Size16M {
+				fileRange[LessThan16M] += 1
+			} else if Size16M <= ino.Size && ino.Size < Size32M {
+				fileRange[LessThan32M] += 1
+			} else if Size32M <= ino.Size && ino.Size < Size64M {
+				fileRange[LessThan64M] += 1
+			} else if Size64M <= ino.Size && ino.Size < Size128M {
+				fileRange[LessThan128M] += 1
+			} else if Size128M <= ino.Size && ino.Size < Size256M {
+				fileRange[LessThan256M] += 1
+			} else {
+				fileRange[BiggerThan256M] += 1
+			}
+		}
+		count++
+		// sleep a while if iter StatsInterval count
+		if count > StatsInterval {
+			count = 0
+			time.Sleep(StatsSleep)
+		}
+	}
+}
+
+func (mp *metaPartition) startFileStats() {
+	checkTicker := time.NewTicker(fileStatsCheckPeriod)
```

Comment:
10s interval may cause cpu cost too much

Reply:
The **fileStatsCheckPeriod** here is mainly used to detect the status change of mp (such as leader to follower). The period for **fileStats** is **fileStatsHandlePeriod** (default value is 15min).